tests/testautorelay.nim

```diff
@@ -77,54 +77,58 @@ suite "Autorelay":
+    await allFutures(switchClient.stop(), switchRelay.stop())
 
   asyncTest "Three relays connections":
-    var state = 0
+    type RelayReservationState = enum
+      Relay1Reserved
+      Relay1AndRelay2Reserved # Although switchClient is connected to rel3, rel3 isn't reserved due to the maximum number of relays set to 2.
+      Relay2UnreservedAndRelay1Reserved
+      Relay1AndRelay3Reserved
+
+    var state = Relay1Reserved
     let
       rel1 = createSwitch(Relay.new())
       rel2 = createSwitch(Relay.new())
       rel3 = createSwitch(Relay.new())
       fut = newFuture[void]()
     relayClient = RelayClient.new()
     proc checkMA(addresses: seq[MultiAddress]) =
-      if state == 0 or state == 2:
-        check:
-          buildRelayMA(rel1, switchClient) in addresses
-          addresses.len() == 1
-        state += 1
-      elif state == 1:
-        check:
-          buildRelayMA(rel1, switchClient) in addresses
-          buildRelayMA(rel2, switchClient) in addresses
-          addresses.len() == 2
-        state += 1
-      elif state == 3:
-        check:
-          buildRelayMA(rel1, switchClient) in addresses
-          buildRelayMA(rel3, switchClient) in addresses
-          addresses.len() == 2
-        state += 1
+      if state == Relay1Reserved or state == Relay2UnreservedAndRelay1Reserved:
+        let relayMAs = buildRelayMA(rel1, switchClient)
+        for relayMA in relayMAs:
+          check: relayMA in addresses
+        if state == Relay1Reserved:
+          state = Relay1AndRelay2Reserved
+        elif state == Relay2UnreservedAndRelay1Reserved:
+          state = Relay1AndRelay3Reserved
+      elif state == Relay1AndRelay2Reserved:
+        let relay1MAs = buildRelayMA(rel1, switchClient)
+        for relayMA in relay1MAs:
+          check: relayMA in addresses
+        let relay2MAs = buildRelayMA(rel2, switchClient)
+        for relayMA in relay2MAs:
+          check: relayMA in addresses
+        state = Relay2UnreservedAndRelay1Reserved
+      elif state == Relay1AndRelay3Reserved:
+        let relay1MAs = buildRelayMA(rel1, switchClient)
+        for relayMA in relay1MAs:
+          check: relayMA in addresses
+        let relay3MAs = buildRelayMA(rel3, switchClient)
+        for relayMA in relay3MAs:
+          check: relayMA in addresses
         fut.complete()
-    let autorelay = AutoRelayService.new(2, relayClient, checkMA, newRng())
+    let autorelay = AutoRelayService.new(maxNumRelays = 2, relayClient, checkMA, newRng())
     switchClient = createSwitch(relayClient, autorelay)
     await allFutures(switchClient.start(), rel1.start(), rel2.start(), rel3.start())
     await switchClient.connect(rel1.peerInfo.peerId, rel1.peerInfo.addrs)
-    await sleepAsync(500.millis)
     await switchClient.connect(rel2.peerInfo.peerId, rel2.peerInfo.addrs)
     await switchClient.connect(rel3.peerInfo.peerId, rel3.peerInfo.addrs)
-    await sleepAsync(500.millis)
     await rel2.stop()
-    await fut.wait(1.seconds)
```

Comment:
I would leave this. Without this await, the test could finish before all the check are done.

Reply:
Actually, all awaits are necessary. It was "working" cause I removed `await fut.wait(1.seconds)` so no check was being done as you said. But I used futures instead, let me know what you think.